File: LeetCode/Binary search/p1011/c++/solution.cpp
class Solution {
public:
    int shipWithinDays(vector<int>& weights, int days) {
            int left = 0, right = 0;
        for(int n:weights){
            left = max(left,n);
            right += n;
        }
        while(left<right){
            int mid = left+(right-left)/2;
            if(weightsCarry(weights, mid)<=days){
                right = mid;
            }
            else{
                left = mid+1;
            }
        }
        return left;
    }
    int weightsCarry(vector<int>& weights, int package){
        int days = 0;
        for(int i=0;i<weights.size();){
            int carry = 0;
            while(i<weights.size()){
                if(package-carry<weights[i]){
                    break;
                }
                carry += weights[i];
                i++;
            }
            days++;
        }
        return days;
    }
};

File: LeetCode/Binary search/p1011/java/solution.java
class Solution {
    public int shipWithinDays(int[] weights, int days) {
        int left = 0,right = 0;
        for(int n:weights) {
            left = Math.max(left,n);
            right += n;
        }
        while(left<=right) {
            int mid = left+(right-left)/2;
            if(spendDays(weights,mid)<=days) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return left;
    }
    int spendDays(int[] weights, int packages) {
        int days = 0;
        for(int i=0;i<weights.length;) {
            int carry = 0;
            while(i<weights.length) {
                if(packages<carry+weights[i]) {
                    break;
                }
                carry += weights[i++];
            }
            days++;
        }
        return days;
    }
}